python/tvm/relay/testing/py_converter.py

```diff
@@ -0,0 +1,367 @@
+        return (self.create_call('TensorValue', [const_expr]), [])
+
+
+    def visit_function(self, func: Expr):
+        # Python's lambdas are very restrictive, so we do "name" inline functions
+        converted_func, func_name = self.convert_func_node('_anon_func', func)
+        return (self.include_var(func_name), [converted_func])
+
+
+    def visit_call(self, call: Expr):
+        '''For calls, we must distinguish between ordinary functions,
+        operators, and constructor calls.'''
+        func = call.op
+        fields, field_defs = self.convert_fields(call.args)
+        if isinstance(func, relay.Op):
+            op_func, op_call = self.create_op_call(func, len(call.args), call.attrs)
+            defs = [op_func] + field_defs
+            return (ast.Call(ast.Name(op_call, ctx=ast.Load()), fields, []), defs)
+
+        if isinstance(func, relay.Constructor):
+            raise Exception('ADTs not supported')
+
+        # ordinary function
+        converted_func, defs = self.visit(func)
+        defs += field_defs
+        return (ast.Call(converted_func, fields, []), defs)
+
+
+    def visit_ref_create(self, ref: Expr):
+        val, defs = self.visit(ref.value)
+        return (self.create_call('RefValue', [val]), defs)
+
+
+    def visit_ref_read(self, read: Expr):
+        ref, defs = self.visit(read.ref)
+        return (ast.Attribute(ref, attr='value', ctx=ast.Load()), defs)
+
+
+    def visit_ref_write(self, write: Expr):
+        '''For writing refs, we wrap the update in a thunk
+        (returning an empty tuple to match Relay's semantics)
+        that we execute at the right time. This ensures such assignments
+        can be properly nested, since assignments are statements
+        in Python but expressions in Relay'''
+        ref, ref_defs = self.visit(write.ref)
+        val, val_defs = self.visit(write.value)
+        thunk_name = self.generate_function_name('_ref_write_thunk')
+        thunk = ast.FunctionDef(thunk_name,
+                                ast.arguments([]),
+                                ref_defs + val_defs + [
+                                    ast.Assign(
+                                        ast.Attribute(ref, attr='value', ctx=ast.Store()),
+                                        val),
+                                    ast.Return(self.create_call('TupleValue', []))
+                                ])
+        return (self.create_call(thunk_name, []), [thunk])
+
+
+    def visit_match(self, _):
+        raise Exception('Pattern matching not yet suported')
+
+
+def to_python(expr: Expr, mod=relay.Module()) -> str:
+    '''Converts the given Relay expression into a Python script.'''
+    converter = PythonConverter(mod)
+    py_ast = converter.convert(expr)
+    return astor.to_source(py_ast)
+
+
+def run_as_python(expr: Expr, mod=relay.Module()):
+    '''Converts the given Relay expression into a Python script and
+    executes it.'''
+    py_ast = to_python_ast(expr, mod)
```

Comment:
Where is `to_python_ast` defined?

Reply:
That was a mistake, it should be calling `to_python` above